src/optimproblem.cpp

```diff
@@ -897,6 +1066,63 @@ void OptimProblem::rollOut(Vec x){
       }
     }
   }
+
+}
+
+void OptimProblem::updateLagrangian(const double prev_mu, const Vec x, Vec lambda) {
+
+  evalF(x, lambda, true);
+
+  Vec lambda_incre;
+  VecCreate(PETSC_COMM_WORLD, &lambda_incre);
+  VecSetSizes(lambda_incre, PETSC_DECIDE, getNstate());
+  VecSetFromOptions(lambda_incre);
+  VecSet(lambda_incre, 0.0);
+  VecAssemblyBegin(lambda_incre);
+  VecAssemblyEnd(lambda_incre);
+
+  /* discontinuity between time segments */
+  Vec disc;
+  VecCreate(PETSC_COMM_WORLD, &disc);
+  VecSetSizes(disc, PETSC_DECIDE, 2*timestepper->mastereq->getDim());
+  VecSetFromOptions(disc);
+
+  for (int iinit = 0; iinit < ninit_local; iinit++) {
+    /* Prepare the initial condition in [rank * ninit_local, ... , (rank+1) * ninit_local - 1] */
+    int iinit_global = mpirank_init * ninit_local + iinit;
+
+    /* Iterate over local time windows */
+    for (int iwindow = 0; iwindow < nwindows_local; iwindow++){
+      if (mpirank_time == mpisize_time-1 && iwindow == nwindows_local-1)
+        continue;
+
+      VecCopy(store_interm_states[iinit][iwindow], disc);
+
+      int index = mpirank_time * nwindows_local + iwindow ;
+      int id = iinit_global*(nwindows-1) + index;
+      Vec xnext;
+      VecGetSubVector(x, IS_interm_states[id], &xnext);
+      VecAXPY(disc, -1.0, xnext);  // finalstate = S(u_{i-1}) - u_i
+
+      /* lag += - prev_mu * ( S(u_{i-1}) - u_i ) */
```

Comment:
If this is the update formula it would be good to have it stated at the top of this function.

Reply:
Yes, we could.

Reply:
Please add the comment

Reply:
Comment added